test/test_quantized.py

```diff
@@ -4,20 +4,83 @@
 import torch.jit
 import numpy as np
 import unittest
-from common_utils import TestCase, run_tests, skipIfNotRegistered
+from common_utils import TEST_WITH_UBSAN, TestCase, run_tests, skipIfNotRegistered
 
 
 def canonical(graph):
     return str(torch._C._jit_pass_canonicalize(graph))
 
 
+def _quantize(x, scale, zero_point):
+    """Quantizes a numpy array."""
+    qx = np.round(x / scale).astype(np.uint8) + zero_point
+    return qx
+
+
+def _dequantize(qx, scale, zero_point):
+    """Dequantizes a numpy array."""
+    x = (qx - zero_point) * scale
+    return x
+
+
+# Make sure we won't have overflows from vpmaddubsw instruction used in fbgemm)
+def avoid_vpmaddubsw_overflow_fc(
+    batch_size, input_channels, output_channels, X, X_min, X_max, W, W_min, W_max
+):
+    for i, j in np.ndindex((batch_size, output_channels)):
+        for k in range(0, input_channels // 2 * 2, 2):
+            x0 = X[i, k] - X_min
+            x1 = X[i, k + 1] - X_min
+            w0 = W[j, k] - 128 - W_min
+            w1 = W[j, k + 1] - 128 - W_min
+            if x0 * w0 + x1 * w1 < -(1 << 15):
+                w1_adjusted = (-(1 << 15) - float(x0) * w0) / x1
+                W[j, k + 1] = int(w1_adjusted) + 128 + W_min
+            elif x0 * w0 + x1 * w1 > (1 << 15) - 1:
+                w1_adjusted = ((1 << 15) - 1 - float(x0) * w0) / x1
+                W[j, k + 1] = int(w1_adjusted) + 128 + W_min
+
+    # Go through the same loop again to double check we don't have any overflow
+    for i, j in np.ndindex((batch_size, output_channels)):
+        for k in range(0, input_channels // 2 * 2, 2):
+            x0 = X[i, k] - X_min
+            x1 = X[i, k + 1] - X_min
+            w0 = W[j, k] - 128 - W_min
+            w1 = W[j, k + 1] - 128 - W_min
+            assert -(1 << 15) <= x0 * w0 + x1 * w1 < (1 << 15)
+
+
+# Reference quantized FC operator
+def qfc_ref(X_q, X_scale, X_zp, W_q, W_scale, W_zp, b_q, Y_scale, Y_zp):
+    row_offsets_ref = X_q.sum(axis=1).astype(np.int32).reshape((-1, 1))
+    col_offsets_ref = W_q.sum(axis=1).astype(np.int32).reshape((1, -1))
+    assert X_q.ndim == 2
+    batch_size, input_channels = X_q.shape
+
+    Prod_XqWq_ref = (
+        np.matmul(X_q.astype(np.int32), W_q.astype(np.int32).T)
+        - W_zp * row_offsets_ref
+        - X_zp * col_offsets_ref
+        + input_channels * X_zp * W_zp
+    )
+
+    Y_q_ref = (
+        (np.round((Prod_XqWq_ref + b_q) * (Y_scale / X_scale / W_scale)) + Y_zp)
+        .clip(0, 255)
+        .astype(np.uint8)
+    )
+    return Y_q_ref
+
+
```

Comment:
What is the purpose of the skip if not registered decorator here?

Reply:
raghuramank100: I think here the decorator means that if the fbgemm-based Caffe2 ops are not linked, then this unit test will be skipped. There should be some dependencies. @zafartahirov Correct me if I am wrong.